### PR TITLE
SNOW-1304211: Pylance throws errors when using copy_options with copy_into_location method

### PR DESCRIPTION
**Pull request by Yoshi-Egawa**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes https://github.com/snowflakedb/snowpark-python/issues/1363

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

* [`src/snowflake/snowpark/dataframe_writer.py`](diffhunk://#diff-30e3ff47797748b9e4b3bf0e50aede5c2a609f6624f51ef83ac2c04802cf17e9L243-R243): The `copy_into_location` method was updated in three instances. The `copy_options` parameter was changed from `Optional[str]` to `Optional[Dict[str, Any]]`. This change allows for a wider range of types to be accepted as copy options, providing more flexibility when calling the method. [[1]](diffhunk://#diff-30e3ff47797748b9e4b3bf0e50aede5c2a609f6624f51ef83ac2c04802cf17e9L243-R243) [[2]](diffhunk://#diff-30e3ff47797748b9e4b3bf0e50aede5c2a609f6624f51ef83ac2c04802cf17e9L259-R259) [[3]](diffhunk://#diff-30e3ff47797748b9e4b3bf0e50aede5c2a609f6624f51ef83ac2c04802cf17e9L274-R274)

### CODE REVIEW

**Comment by Yoshi-Egawa** [comment]
I have read the CLA Document and I hereby sign the CLA

File: src/snowflake/snowpark/dataframe_writer.py
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/dataframe_writer.py:242]:
```suggestion
        block: Literal[True],
```

File: src/snowflake/snowpark/dataframe_writer.py
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/dataframe_writer.py:258]:
```suggestion
        block: Literal[False],
```

**Review by sfc-gh-stan — APPROVED** [review]
Thanks for your contribution!
The changes LGTM, I suggested 2 additional type hint fixes.

**Comment by Yoshi-Egawa** [comment]
@sfc-gh-stan 
> I suggested 2 additional type hint fixes.

Thank you for your suggestion. I'd like to propose a slightly different fix for the blocks type.

The current suggestion sets block: `Literal[True]`. However, I believe it should be block: `bool` to maintain type consistency with the `_internal_collect_with_tag_no_telemetry` function.

The blocks argument in the `_internal_collect_with_tag_no_telemetry` function is used as a bool type. Therefore, setting the blocks type to bool in the `copy_into_location` function as well will ensure type consistency.


`src/snowflake/snowpark/dataframe.py`
```python
    def _internal_collect_with_tag_no_telemetry(
        self,
        *,
        statement_params: Optional[Dict[str, str]] = None,
        block: bool = True,  # <-- here
        data_type: _AsyncResultType = _AsyncResultType.ROW,
        log_on_exception: bool = False,
        case_sensitive: bool = True,
    ) -> Union[List[Row], AsyncJob]:
        # When executing a DataFrame in any method of snowpark (either public or private),
        # we should always call this method instead of collect(), to make sure the
        # query tag is set properly.
        return self._session._conn.execute(
            self._plan,
            block=block,
            data_type=data_type,
            _statement_params=create_or_update_statement_params_with_query_tag(
                statement_params or self._statement_params,
                self._session.query_tag,
                SKIP_LEVELS_THREE,
            ),
            log_on_exception=log_on_exception,
            case_sensitive=case_sensitive,
        )

    _internal_collect_with_tag = df_collect_api_telemetry(
        _internal_collect_with_tag_no_telemetry
    )
```

**Comment by Yoshi-Egawa** [comment]
@sfc-gh-stan 
Thank you for your suggestion. I have implemented your suggestion with a slight modification.

I have added a default value to the blocks keyword argument. This is because keyword arguments are optional, and not providing a value would result in an error.

**Review by sfc-gh-stan — APPROVED** [review]

**Review by sfc-gh-jrose — APPROVED** [review]

**Comment by sfc-gh-stan** [comment]
Tests kicked off by #1377, will merge after merge gates pass.

**Pull request merged**
